endpoints/cookie_sync.go

```diff
@@ -192,6 +197,18 @@ func (req *cookieSyncRequest) filterForGDPR(permissions gdpr.Permissions) {
 	}
 }
 
+// filterToLimit will enforce a max limit on cookiesyncs supplied, picking a random subset of syncs to get to the limit if over.
+func (req *cookieSyncRequest) filterToLimit() {
+	if req.Limit <= 0 {
+		return
+	}
+	for len(req.Bidders) > req.Limit {
+		cut := rand.Intn(len(req.Bidders))
+		req.Bidders = append(req.Bidders[:cut], req.Bidders[cut+1:]...)
```

Comment:
Maybe there is a more efficient way of taking randomly the `req.Limit` elements leveraging `rand.Perm` returning a random permutation of the numbers 0 through N (not including N). It will avoid creating new slices and the `O(n^2)` operation.

```
biddersToKeep = make([]req.Bidders, req.Limit)
for n, idx := range rand.Perm(len(req.Bidders)) {
      if n == req.Limit {
          break
      }
      biddersToKeep[n] = req.Bidders[idx]
  }

req.Bidders = biddersToKeep
```

Reply:
There is maybe also a way to shuffle the slice in place index by index until we reached the limit.
Something like
```
randIndex = rand.Intn(len(req.Bidders))
vals[n], vals[randIndex] = vals[randIndex], vals[n]
```

Reply:
yeah... you're quickly converging on the Fisher-Yates shuffle: https://en.wikipedia.org/wiki/Fisher%E2%80%93Yates_shuffle#The_modern_algorithm

Run that over the first `Limit` elements only, and then `req.Bidders = req.Bidders[:limit]` to set the slice. Either way, there are definitely ways to do this in O(limit) time and O(1) space.

Reply:
Though if on average we discard more than 2/3rds of the bidders, then starting from the front would be more efficient.

Reply:
We probably would discard more than 2/3s of the bidders in most cases... we're already at 20 bidders, which would be a sync limit of ~7. That's already quite a few requests for a page to make.

I wouldn't worry about it too much though. It's fine as is, or fine if you want to change it.